File: scripts/session-cost.ts
#!/usr/bin/env node

// Prices one session's transcript and writes its row under
// `.claude/costs/sessions/`. The row is rewritten from the whole file each run
// rather than appended to, which is what lets a run pick up anything the
// previous one was too early to see. `--at-stop` is the Stop hook's: the turn is
// over, so the transcript should end on its `end_turn`. `--out` is the hook's
// too: it writes the row there instead of into place, since committing it is the
// hook's job.
//
//   node scripts/session-cost.ts --transcript <path> [--session-id <id>] [--row-path] [--at-stop] [--out <path>]

/* eslint-disable no-console -- stdout is this script's interface: the row's
   path for the hook that calls it, a one-line summary for a person running it
   by hand. */

import { readdirSync, readFileSync, writeFileSync } from 'node:fs';
import path from 'node:path';

import { flag, given } from './lib/argv.ts';
import { rowText } from './lib/cost-rows.ts';
import {
  isUnwrittenTail,
  parsePrices,
  parseSessionCost,
  type SessionCost,
  summariseTranscript,
} from './lib/session-cost.ts';
import { writeAtomic } from './lib/write-atomic.ts';

const root = process.env['CLAUDE_PROJECT_DIR'] ?? process.cwd();

const transcript = flag('transcript');
if (transcript === undefined) {
  console.error('session-cost: --transcript <path> is required');
  process.exit(2);
}

// The month a session is filed under is the month it started, so a session
// running across midnight on the last of the month stays in one file.
const monthOf = (cost: SessionCost): string =>
  (cost.firstResponseAt ?? new Date().toISOString()).slice(0, 7);

// A subagent's responses are billed to this session and written to their own
// file under `<transcript>/subagents/`, so the directory is read rather than
// assumed empty. A session that spawned none has no directory at all.
const subagentsOf = (main: string): string[] => {
  const dir = path.join(
    path.dirname(main),
    path.basename(main, '.jsonl'),
    'subagents',
  );
  try {
    return readdirSync(dir)
      .filter((name) => name.endsWith('.jsonl'))
      .map((name) => readFileSync(path.join(dir, name), 'utf8'));
  } catch {
    return [];
  }
};

// An unwritten-tail warning is what no run can recompute from the transcript,
// so a rewrite reads it back from the last one. An unreadable row is treated as
// no row: the point is to keep the warning, never to fail a write over one.
const previous = (row: string): SessionCost | undefined => {
  try {
    return parseSessionCost(readFileSync(row, 'utf8'));
  } catch {
    return undefined;
  }
};

const prices = parsePrices(
  readFileSync(path.join(root, '.claude/costs/prices.json'), 'utf8'),
);
const cost = summariseTranscript(
  {
    main: readFileSync(transcript, 'utf8'),
    subagents: subagentsOf(transcript),
  },
  prices,
  flag('session-id') ?? path.basename(transcript, '.jsonl'),
  given('at-stop'),
);

const out = path.join(
  root,
  '.claude/costs/sessions',
  monthOf(cost),
  `${cost.sessionId}.json`,
);
const before = previous(out);
const carried = (before?.warnings ?? []).filter(
  (warning) => isUnwrittenTail(warning) && !cost.warnings.includes(warning),
);
const row: SessionCost = {
  ...cost,
  warnings: [...carried, ...cost.warnings],
};
const staged = flag('out');
if (staged === undefined) writeAtomic(root, out, rowText(row));
else writeFileSync(staged, rowText(row));

console.log(
  given('row-path')
    ? out
    : `session-cost: ${row.total.responses} responses, $${row.total.costUsd.toFixed(4)} → ${out}`,
);

File: .claude/costs/session_cost.py
#!/usr/bin/env python3
"""Price one session's transcript and write its row under `sessions/`.

Usage:
  python3 .claude/costs/session_cost.py --transcript <path> [--session-id <id>] [--events <path>] [--row-path] [--at-stop] [--out <path>]

The row is rewritten from the whole file each run rather than appended to, which
is what lets a run pick up anything the previous one was too early to see.
`--at-stop` is the Stop hook's: the turn is over, so the transcript should end on
its `end_turn`. `--out` is the hook's too: it writes the row there instead of into
place, since committing it is the hook's job. Stdout is the interface: the row's
path under `--row-path`, for `hooks/stop-session-cost.sh`; a one-line summary
otherwise, for a person running it by hand. `--events` names the session's
telemetry file where it is not the receiver's `tmp/telemetry/<session-id>.jsonl`;
a session with none is priced from its transcript alone.

Paths resolve from this file's own location, so it runs from any working
directory. Stdlib only — Python 3.9+.
"""

from __future__ import annotations

import argparse
import json
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, List, Optional

from lib.billed import Event, parse_events
from lib.pricing import (
    TranscriptSources,
    is_unwritten_tail,
    parse_prices,
    summarise_transcript,
)
from lib.rows import ROOT, SessionCost, parse_session_cost, row_text, write_atomic
from lib.shape import ShapeError

COSTS = Path(__file__).resolve().parent


def subagents_of(main: Path) -> List[str]:
    """A subagent's responses are billed to this session and written to their own
    file under `<transcript>/subagents/`, so the directory is read rather than
    assumed empty. A session that spawned none has no directory at all."""
    directory = main.parent / main.stem / "subagents"
    if not directory.is_dir():
        return []
    return [path.read_text(encoding="utf-8") for path in sorted(directory.glob("*.jsonl"))]


def previous(row: Path) -> Optional[SessionCost]:
    """An unwritten-tail warning is what no run can recompute from the
    transcript, so a rewrite reads it back from the last one. An unreadable row
    loses it rather than failing the write — the rewrite is what repairs it —
    and says so."""
    if not row.exists():
        return None
    try:
        return parse_session_cost(row.read_text(encoding="utf-8"), str(row))
    except (ShapeError, json.JSONDecodeError) as error:
        print(
            f"session-cost: {error}; rewriting the row from the transcript alone",
            file=sys.stderr,
        )
        return None


def events_at(path: Path) -> Optional[Dict[str, Event]]:
    if not path.exists():
        return None
    return parse_events(path.read_text(encoding="utf-8"), str(path))


def month_of(cost: SessionCost) -> str:
    """The month a session is filed under is the month it started, so a session
    running across midnight on the last of the month stays in one file."""
    started = cost.first_response_at or datetime.now(timezone.utc).isoformat()
    return started[:7]


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__.split("\n\n")[0])
    parser.add_argument("--transcript", required=True, type=Path)
    parser.add_argument("--session-id")
    parser.add_argument("--events", type=Path)
    parser.add_argument("--row-path", action="store_true")
    parser.add_argument("--at-stop", action="store_true")
    parser.add_argument("--out", type=Path)
    args = parser.parse_args()

    transcript: Path = args.transcript
    session_id: str = args.session_id or transcript.stem
    events = events_at(args.events or ROOT / "tmp" / "telemetry" / f"{session_id}.jsonl")
    prices = parse_prices((COSTS / "prices.json").read_text(encoding="utf-8"))
    cost = summarise_transcript(
        TranscriptSources(
            main=transcript.read_text(encoding="utf-8"),
            subagents=subagents_of(transcript),
        ),
        prices,
        session_id,
        at_stop=args.at_stop,
        events=events,
    )

    out = COSTS / "sessions" / month_of(cost) / f"{cost.session_id}.json"
    before = previous(out)
    if before is not None:
        carried = [w for w in before.warnings if is_unwritten_tail(w) and w not in cost.warnings]
        cost.warnings = carried + cost.warnings
    row = row_text(cost)
    if args.out is not None:
        args.out.write_text(row, encoding="utf-8")
    else:
        write_atomic(out, row)

    if args.row_path:
        print(out)
    else:
        print(f"session-cost: {cost.total.responses} responses, ${cost.total.cost_usd:.4f} → {out}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
